### DISCUSSION TITLE
what is purpose of idle vs prepare handles?

### DISCUSSION THREAD

**Discussion by dominictarr** [libuv/libuv#3438] [Q&A]
Except that idle handles are called before prepare handles, both have the same behavior and function signatures.
they seem like they both do exactly the same thing, any feature implemented with idle could instead be done with prepare, and vice versa

Actually, I'm more confused because I can't figure out where `uv_idle_init` (etc) methods are actually defined? `grep uv_idle_init -R ./src` is not returning a definition.

**Comment by cjihrig** [comment]
I'm not sure if this answers your question, but according to the idle handle docs:

> The notable difference with prepare handles is that when there are active idle handles, the loop will perform a zero timeout poll instead of blocking for i/o.

The functions you're looking for are defined in a macro named `UV_LOOP_WATCHER_DEFINE` in `src/{unix,win}/loop-watcher.c`.

**Reply by dominictarr** [reply]
right, so idle handles will run much more hotly than prepare timers... and actually cause your program to wait in the idle callback loop (constantly calling idle callbacks then checking if io has happened) rather than waiting in the io loop, (calling a prepare callback then waiting until io happens)

**Reply by cjihrig** [reply]
Yea, I believe that is correct. There was even some discussion about renaming and potentially even removing idle handles for [libuv v2 back in 2016](https://github.com/libuv/libuv/issues/1035).
